### PR TITLE
[Chat] Fix lora

### PR DESCRIPTION
**Pull request by YeAnbang**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: applications/ColossalChat/tests/test_lora.py [hunk 1]
```diff
@@ -65,5 +65,50 @@ def test_overfit():
...
+    linear = nn.Linear(5, 10)
+    linear.weight.data = weight
+    x = torch.randn(10, 5)
+    out_linear = linear(x)
+
+    # lora linear Pissa
+    linear.weight.data = weight
+    lora_linear = LoraLinear(linear.weight, linear.bias, r=2, lora_initialization_method="PiSSA")
+    out_lora = lora_linear(x)
+    assert torch.allclose(out_linear, out_lora, atol=1e-5)
```
Comment (TongLi3701) [inline applications/ColossalChat/tests/test_lora.py:80]: Normal we should also check `rtol`

**Review by TongLi3701 — CHANGES_REQUESTED** [review]
Thanks Anbang.

Just left some comments. 

Normally we should also support Lora for RLHF methods. Please also add tests or instructions for them. Thanks.

File: applications/ColossalChat/coati/models/lora.py [hunk 2]
```diff
@@ -5,161 +5,365 @@
...
 
+    def reset_parameters(self):
+        if hasattr(self, "lora_A"):
+            if self.lora_initialization_method == "kaiming_uniform" or self.weight.size() != (
+                self.out_features,
+                self.in_features,
+            ):
+                # Initialize A with the default values for nn.Linear and set B to zero.
+                nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5))
+                nn.init.zeros_(self.lora_B)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/models/lora.py:77]: no need nn.Parameter has require_grad=True by default

File: applications/ColossalChat/coati/models/lora.py [hunk 3]
```diff
@@ -5,161 +5,365 @@
...
+    # make all parameter not trainable, if lora_train_bias is "all", set bias to trainable
+    for name, p in module.named_parameters():
+        p.requires_grad = False
+        if "bias" in name and lora_config.lora_train_bias == "all":
+            p.requires_grad = True
+    _convert_to_lora_recursively(module, "", lora_config)
+    trainable_parameter_size = 0
+    for name, p in module.named_parameters():
+        if p.requires_grad == True:
+            trainable_parameter_size += p.numel()
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/models/lora.py:366]: The trainable parameter size captures the trainable parameters after converting to lora model, so only lora_A and lora_Bs count.

File: applications/ColossalChat/coati/models/lora.py [hunk 4]
```diff
@@ -5,161 +5,365 @@
...
-        fan_in_fan_out: bool = False,
+        lora_initialization_method: str = "kaiming_uniform",
     ):
-        nn.Module.__init__(self)
-        lora.LoRALayer.__init__(self, r=r, lora_alpha=lora_alpha, lora_dropout=lora_dropout, merge_weights=False)
+        super().__init__(
+            r=r, lora_alpha=lora_alpha, lora_dropout=lora_dropout, lora_initialization_method=lora_initialization_method
+        )
+
+        if weight is not None:
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/models/lora.py:141]: removed

File: applications/ColossalChat/coati/models/lora.py [hunk 5]
```diff
@@ -5,161 +5,365 @@
...
-            self.weight.data = self.weight.data.T
 
-    def reset_parameters(self):
-        if hasattr(self, "lora_A"):
-            # Initialize A with the default values for nn.Linear and set B to zero.
-            nn.init.kaiming_uniform_(self.lora_A, a=math.sqrt(5))
-            nn.init.zeros_(self.lora_B)
+        # reset parameters
+        nn.init.zeros_(self.lora_A)
+        nn.init.normal_(self.lora_B)
```
[the remark this answers is not in the document]
Comment (YeAnbang) [reply applications/ColossalChat/coati/models/lora.py:220]: yes

**Review by TongLi3701 — APPROVED** [review]
Thanks Anbang.

**Pull request merged**